### PR TITLE
#22517 fix invalid redirect location

### PR DESCRIPTION
**Pull request by funivan**
| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #22517
| License       | MIT
| Doc PR        | n/a

### CODE REVIEW

**Comment by issei-m** [comment]
A bug fix should be meged oldest maintained branch (i.e 2.7)

**Comment by funivan** [comment]
@issei-m What i need to do? Just change branch in my PR or checkout from the 2.7 branch and create new PR?

**Comment by stof** [comment]
The phpdoc explicit says it expects a `string`. A castable object is not a string. So this is not a bugfix. It is a new feature

**Comment by funivan** [comment]
@stof ok. So to prevent this situations we should fire error if parameter is invalid?
Or I can change this PR to New feature instead of "bugfix"

File: src/Symfony/Component/HttpFoundation/RedirectResponse.php [hunk 1]
```diff
@@ -36,7 +36,7 @@ public function __construct($url, $status = 302, $headers = array())
     {
         parent::__construct('', $status, $headers);
 
-        $this->setTargetUrl($url);
+        $this->setTargetUrl((string) $url);
```
Comment (xabbuh) [inline src/Symfony/Component/HttpFoundation/RedirectResponse.php]: The docblock already states that `$url` must be a string.

**Review by xabbuh — CHANGES_REQUESTED** [review]
We can improve the docblock of the `create()` method instead to make it clear that the passed `$url` must be a string (currently the docblock inherits from the parent `Response` class).

**Comment by funivan** [comment]
@xabbuh Update phpdoc for the `create` method. But tests failed in the `Lock` component. Can you restart it? Thank you.

**Comment by funivan** [comment]
So here is second try. Deny any invalid url to be pass to the `RedirectResponse`
We can pass only `not empty` `string`
Documentation of `create` method is fixed.

**Comment by stof** [comment]
@funivan we don't add exceptions for all cases where scalar params are wrong. This would require adding lots of code in Symfony, with overhead both for runtime and maintenance.

When you pass arguments which don't respect the documented contract of an API, you enter unknown land. The behavior is undefined. It may work by luck due to an implementation details, but there is no guarantee about it (and your luck may disappear even in a patch release if the implementation detail making it work changes).
So if you want to represent your URL as a value object, cast it as string *before* passing it to the RedirectResponse, to give it a string.

File: src/Symfony/Component/HttpFoundation/RedirectResponse.php (hunk 1 above)
Comment (funivan) [inline src/Symfony/Component/HttpFoundation/RedirectResponse.php]: Done. I removed this type cast

**Comment by funivan** [comment]
@stof I get your point. So what about "new feature" 
It can nice to have automatic type casting? I can see that `Response` has already this feature
https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpFoundation/Response.php#L408:L412

File: src/Symfony/Component/HttpFoundation/RedirectResponse.php [hunk 2]
```diff
@@ -76,6 +83,10 @@ public function getTargetUrl()
      */
     public function setTargetUrl($url)
     {
+        if (!is_string($url)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/RedirectResponse.php]: If we were to be on PHP7, the cast to string would be done by the engine. What about just doing: `$url = (string) $url` ?

Comment (funivan) [inline src/Symfony/Component/HttpFoundation/RedirectResponse.php]: 100% agree and I made it at first https://github.com/funivan/symfony/commit/79038731461e2d694d4c96be0cd8db108ddae343

But then according to the comments I change this behaviour. 
If cast to string will be better choice, should I change phpdoc comment?

Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/RedirectResponse.php]: the doc comment is fine to me, we only accept string - and we work around the fact that the engine cannot do any cast for us until php7 by doing the cast. But the cast should be done inside the function, not when calling it as you did in that commit.
Note that I'm still not sure we should do this here. Can't you do the cast on *your* side?

Comment (funivan) [inline src/Symfony/Component/HttpFoundation/RedirectResponse.php]: Cast on my side works fine. If we do not cast object then response contains valid body and invalid location header. This is the main argument to do this inside the function.

**Comment by nicolas-grekas** [comment]
To help move this forward, I think the docblock change is nice and shoud be applied on 2.7, but the added exception is not worth it and is just an unsupported area - meaning #22517 should be closed as "bogus".
@funivan OK to close and submit the docblock on 2.7?

**Comment by funivan** [comment]
@nicolas-grekas ok. Totally agree.

**Pull request closed without merging**